Implement dietline history on mount-shell (#8806).

davidpolverari opened the pull request:
Gives minimal dietline history support inside mount-shell. There is
still a need to implement autocompletion.

radare commented:
RFS doesnt depends on RCons, so this shouldn't build.

It will be better if we move the logic of reading from stdin to a callback (Maybe RConsBind ?) and set all the autocompletion logic outside rfs, maybe in libr/core/cfs.c or cmd_fs.c ?

radare commented:
ping

davidpolverari commented:
I’m still trying to figure out a way to implement this correctly, by studying the way things are structured in r2. But it’s taking some time, because I’m not familiar with the codebase. What if we dropped mount shell and implemented platform-agnostic utils (think r2cat, r2ls) to replace ms commands? Wouldn’t it simplify this issue?

radare commented:
as shell programs? dont think so, you need to keep a state of the RFS mountpoints in a single session. r2 have its own implementations of `ls`, `cp`… but using r_sys apis, not r_fs ones. all that logic is implemented under `m` commands.

> On 1 Mar 2018, at 11:24, davidpolverari <notifications@github.com> wrote:
> 
> I’m still trying to figure out a way to implement this correctly, by studying the way things are structured in r2. But it’s taking some time, because I’m not familiar with the codebase. What if we dropped mount shell and implemented platform-agnostic utils (think r2cat, r2ls) to replace ms commands? Wouldn’t it simplify this issue?
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub <https://github.com/radare/radare2/pull/9534#issuecomment-369547097>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AA3-lnhsX6lw12EobNcbK1GoqTgfDwFEks5tZ8xXgaJpZM4SWA4g>.
>

radare commented:
can you cleanup this PR? just squash your commits and rebase it on top of master, otherwise its a hell to review

davidpolverari commented:
I squashed the commits and rebased, but I still haven't been able to move the logic outside fs.c, neither implement autocompletion.

radare commented:
I think RFs should provide an RFSShell struct to work with it:
```c
RFSShell * shell = r_fs_shell_new(fs);
r_fs_shell_run(shell, "cd /root");
r_fs_shell_run(shell, "cat README.md");
r_fs_shell_free(shell);
```

this way we can initialize some callbacks in RFSShell like readline() or autocomplete() and use that from RCore to move the logic of the repl from fs to core.

davidpolverari commented:
I wonder if you could review my last commit, as I managed to move things out of fs and into cmd_mount.c. As it stands, it already gives history support, but still no autocomplete. I think it's now more in line with what you suggested, and I wanted to know if I'm on the right track.

radare commented:
fuck forgot to send that review.

davidpolverari commented on libr/fs/fs.c:
```diff
@@ -700,13 +702,15 @@ R_API int r_fs_prompt(RFS* fs, const char* root) {
+
+		if (!ptr) break;
+
+		strncpy (buf, ptr, PROMT_PATH_BUFSIZE);
```
In this case, I think it's useful to keep that copy, because later in other places there are copies made considering sizeof(buf) that wouldn't work with sizeof(ptr), as ptr is const char*. But will fix it too.

davidpolverari commented:
I applied the fixes to the commit. Thanks.

radare commented on libr/fs/fs.c:
```diff
@@ -700,13 +702,17 @@ R_API int r_fs_prompt(RFS* fs, const char* root) {
 		}
-		buf[strlen (buf) - 1] = '\0';
+
+		strncpy (buf, ptr, sizeof (buf) - 1);
```
Strncpy doesnt null terminate. Use the r_str_ncpy from r_util instead

radare commented on libr/fs/fs.c:
```diff
@@ -700,13 +702,17 @@ R_API int r_fs_prompt(RFS* fs, const char* root) {
-		fflush (stdout);
-		fgets (buf, sizeof (buf) - 1, stdin);
-		if (feof (stdin)) {
+		strncpy (prompt, path, sizeof (prompt) - 1);
```
Better use snprintf and chop path if longer than PROMT_PATH..

radare commented on libr/fs/fs.c:
```diff
@@ -700,13 +702,17 @@ R_API int r_fs_prompt(RFS* fs, const char* root) {
-		fgets (buf, sizeof (buf) - 1, stdin);
-		if (feof (stdin)) {
+		strncpy (prompt, path, sizeof (prompt) - 1);
+		strcat (prompt, "> ");
```
This will overflow if the previous strncpy fills the buffer. And its doing an unnecessary strlen on the string copied in the line above. Just do snprintf

radare commented on libr/fs/fs.c:
```diff
@@ -700,13 +702,17 @@ R_API int r_fs_prompt(RFS* fs, const char* root) {
-		if (feof (stdin)) {
+		strncpy (prompt, path, sizeof (prompt) - 1);
+		strcat (prompt, "> ");
+		shell->set_prompt (prompt);
```
Maybe add nullchks before calling?

radare commented on libr/fs/fs.c:
```diff
@@ -700,13 +702,17 @@ R_API int r_fs_prompt(RFS* fs, const char* root) {
+		strncpy (prompt, path, sizeof (prompt) - 1);
+		strcat (prompt, "> ");
+		shell->set_prompt (prompt);
+		ptr = shell->readline ();
```
Here too

radare commented on libr/fs/fs.c:
```diff
@@ -700,13 +702,17 @@ R_API int r_fs_prompt(RFS* fs, const char* root) {
+		strcat (prompt, "> ");
+		shell->set_prompt (prompt);
+		ptr = shell->readline ();
+		shell->hist_add (ptr);
```
And there

davidpolverari commented:
I have just fixed the issues. Thanks again for the review and sorry about that many mistakes.

radare commented on libr/fs/fs.c:
```diff
@@ -700,13 +702,22 @@ R_API int r_fs_prompt(RFS* fs, const char* root) {
-		if (feof (stdin)) {
+		snprintf (prompt, sizeof (prompt), "[%.*s]> ", sizeof (prompt) - 5, 
+		  path);
+		if (!shell->set_prompt ||
```
Its ok if we cant histadd or setprompt. And readline can be defaulted to fgets, dont be that strict, because this repl should work even with shell as null

davidpolverari replied:
I'm not sure I understood what has to be made in this case. Should I use an "if" to check if shell is NULL and then use fgets instead, at the same time I don't call hist_add?

radare commented:
Sont worryabout the misstakes and comments. There are always things to improve everywhere. Thanks for caring

radare commented:
Having tab is the important thing of having dietline in here, could you give it a try in this pr or prefer to get it merged after fixing the usage with shell==NULL and then continue in another pr?

> On 18 Mar 2018, at 02:09, davidpolverari <notifications@github.com> wrote:
> 
> I have just fixed the issues. Thanks again for the review and sorry about that many mistakes.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

davidpolverari commented:
I prefer to get this merged after I fix the shell==NULL issue, and continue tab completion in other PR, because I will have understand better how completion works in r2 before starting it, and because I have some work stuff to do before I can concentrate on it.

davidpolverari commented:
This last commit implements what I understood from the shell == NULL issue you pointed. But I think mine still isn't the best implementation in this case, as this is making the check every time the loop iterates. I didn't squash the commits yet, as I think this one won't be the final implementation of the solution you asked, but maybe you can tell if this is the approach you were thinking about.

Besides, I didn't implement a function to malloc a RFSShell and return a pointer to it (a _new() function), as I thought that only one instance of the structure would be needed in order to pass the pointers to the r_cons functions to be called from within fs.c. This way, the RFSShell struct used on cmd_mount.c as a parameter when calling r_fs_shell_prompt() is allocated on the stack and won't ever point to NULL, even when not initialized. Thus, the only way to trigger the code I added to this commit is to call r_fs_shell_prompt() with the first argument as NULL. So I still didn't realize the benefits of this check.

radare commented:
Ive merged and fixed my comments, check the next commit. Would you like to implement the autocompletion?

davidpolverari commented:
Yes, I would like to implement it. I just wanted to warn you that it might take some time, because I'll have to understand how it is already done in r2, in order to do it right.
